mechanics/tools/gdocs-migrate/forest.go

```diff
@@ -0,0 +1,60 @@
+package main
+
+type File struct {
+	id       string
+	name     string
+	parentId string
+	isFolder bool
+}
+
+type Forest struct {
+	self     map[string]File
+	children map[string][]File
+	leaves   []File
+}
+
+func NewForest() Forest {
+	return Forest{
+		self:     make(map[string]File),
+		children: make(map[string][]File),
+	}
+}
+
+func (fo *Forest) Add(f File) {
+	fo.self[f.id] = f
+
+	fo.children[f.parentId] = append(fo.children[f.parentId], f)
+	if !f.isFolder {
+		fo.leaves = append(fo.leaves, f)
+	}
+}
+
+func (fo *Forest) GetFiles() []File {
+	return fo.leaves
+}
+
+func (fo *Forest) Size() int {
+	return len(fo.self)
+}
+
+func (fo *Forest) GetPath(f File) []string {
+	parents := []string{}
```

Comment:
in theory you can just store it in the file itself when you create the file, then it's just `return path`

Reply:
we don't have a guarantee that we receive the parent metadata before the child.

Reply:
How do you build path if you don't have parent MD?

Reply:
`GetPath` is only called after all the folder data is collected, so we'll have the data then, but we won't necessarily have it during `Add`.

Reply:
But Victor's point is that we're only passing in `i.Parents[0]`, but the API will give us all the ancestors.

Reply:
the Parents array give us all the direct Parents (apparently one File may belong to a few Parent -- fortunately none of our File is that way).

that also means Parents[0] is all that we get... 

I've updated doc that we should call these after all data has been loaded.